Slot mappings documentation update (#10011)
RasaHQ/rasa: docs/docs/slot-validation-actions.mdx

---
id: slot-validation-actions
sidebar_label: Slot Validation Actions
title: Slot Validation Actions
abstract: A slot validation action is a special type of custom action, designed to handle custom extraction and/or validation of slot values. This can be used to validate slots with predefined mappings or extract slots with custom mappings.
---

> Rasa SDK provides three base classes that you can use to inherit from in order to define your custom slot extraction and
> validation code:

Reviewer: ```suggestion
Rasa SDK provides three base classes that you can inherit from in order to implement custom slot extraction and
validation logic:
```

## `Action` class

> You can use this class to build a [custom action](./custom-actions.mdx) that is limited to returning only
> `SlotSet` and `BotUttered` events. Any other event type will be filtered out by the default action

Reviewer: ```suggestion
You can use this class to create a regular [custom action](./custom-actions.mdx). The base class will not prevent you from returning arbitrary events, however a slot validation action should only return
`SlotSet` and `BotUttered` events. Any other event type will be filtered out by the default action
```

[`action_extract_slots`](./default-actions.mdx#action_extract_slots).

> The name of the custom action will have to be specified in the `action` key of the

Reviewer: ```suggestion
The name of the custom action must be specified in the `action` key of the relevant
```

[custom slot mapping](./domain.mdx#custom-slot-mappings) in the domain.
Note that the action name must be listed in the domain `actions` section too.

## `ValidationAction` class

> You can extend the `ValidationAction` in the Rasa SDK to define custom extraction and / or validation of slots that

Reviewer: ```suggestion
You can extend the `ValidationAction` class in the Rasa SDK to define custom extraction and / or validation of slots that
```

can be set or updated outside of a form context.
With this option, you do not need to specify the `action` key in the [custom slot mapping](./domain.mdx#custom-slot-mappings).

Note that you do not need to implement the `name` method in the custom action extending `ValidationAction`, since this

> is already provided: `action_validate_slot_mappings`. However this name must be added to the domain `actions` list.

Reviewer: What happens if a user overrides the name? It never gets called right? If so worth mentioning here

Author: Yes, that's correct, the custom validation action will never run, I'll add a comment.

:::note
Slots mapped to a form in the mapping `conditions` will be ignored by `ValidationAction` unless you override
the `required_slots` method to include the slots you want to extract and / or validate.
:::

### Validation of Slots with Predefined Mappings

> You must write functions named `validate_<slot_name>` for every predefined slot that you need validated.

Reviewer: ```suggestion
To validate slots with a predefined mapping, you must write functions named `validate_<slot_name>`.
```

In the following example, the value for slot `location` is capitalized only if the extracted value is of type string:

```python
from typing import Text, Any, Dict

from rasa_sdk import Tracker, ValidationAction
from rasa_sdk.executor import CollectingDispatcher
from rasa_sdk.types import DomainDict

class ValidatePredefinedSlots(ValidationAction):
    def validate_location(
        self,
        slot_value: Any,
        dispatcher: CollectingDispatcher,
        tracker: Tracker,
        domain: DomainDict,
    ) -> Dict[Text, Any]:
        """Validate location value."""
        if isinstance(slot_value, str):
            # validation succeeded, capitalize the value of the "location" slot
            return {"location": slot_value.capitalize()}
        else:
            # validation failed, set this slot to None
            return {"location": None}
```

### Extraction of Custom Slot Mappings

In order to define custom extraction code, you must write an `extract_<slot_name>` method for every slot with a

> custom slot mapping.

Reviewer: What happens if the user specifies a custom action in the `action` key of a custom slot mapping, and then provides a validation method here? Does that work?

Author: Yup, this _should_ work, however there's a reported bug blocking this from working fully (by that I mean that the ValidationAction fails to run for the extracted slot value by the custom action from the `action` key), but should be fixed before the release 🤞🏼

Reviewer: Got it, are we sure enough about that to update this section accordingly? Else that bug should get an additional "Change docs when you fix this" note

Author: I actually fixed the bug that I thought was causing the problem yesterday, but the problem persists so I created this issue in `rasa-sdk` repo: https://github.com/RasaHQ/rasa-sdk/issues/593
To be earnest, I don't know if this will be prioritised before the release - mainly because I think it should not be a recommended route of extraction and validation (I think it's better design for `actions.py` to have extraction and validation for one slot in one place, either the custom action or the `ValidationAction`; if the custom action is supposed to return a `SlotSet` event it should have the validated value already). I can add the note to change docs to its definition of done.

The following example shows the implementation of a custom action that extracts the slot `count_of_insults` to keep
track of the user's attitude.

```python
from typing import Dict, Text, Any

from rasa_sdk import Tracker
from rasa_sdk.executor import CollectingDispatcher
from rasa_sdk.forms import ValidationAction

class ValidateCustomSlotMappings(ValidationAction):
    async def extract_count_of_insults(
        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
    ) -> Dict[Text, Any]:
        intent_of_last_user_message = tracker.get_intent_of_latest_message()
        current_count_of_insults = tracker.get_slot("count_of_insults")
        if intent_of_last_user_message == "insult":
           current_count_of_insults += 1

        return {"count_of_insults": current_count_of_insults}
```

## `FormValidationAction` class

> If you prefer that certain custom slot mappings should only be set within the context of a form,

Reviewer: Bit confused by this condition phrasing because `conditions` already allow you to limit applicability of any mapping to a form. 
Will all validation actions run regardless of conditions? If so, this could read:
```suggestion
If certain custom slot mappings should only be validated within the context of a form,
```

Author: The mapping `conditions` actually have no effect if a custom extraction was defined in an action extending `ValidationAction` instead of `FormValidationAction`. That's why I tried to emphasise that if there are any custom slot mappings meant to be both extracted and validated only within a form, better use `FormValidationAction`.

---
id: slot-validation-actions
sidebar_label: Slot Validation Actions
title: Slot Validation Actions
abstract: A slot validation action is a special type of custom action, designed to handle custom extraction and/or validation of slot values. This can be used to validate slots with predefined mappings or extract slots with custom mappings.
---

Rasa SDK provides three base classes that you can inherit from in order to implement custom slot extraction and
validation logic:

## `Action` class

You can use this class to create a regular [custom action](./custom-actions.mdx). The base class will not prevent you
from returning arbitrary events, however a slot validation action should only return `SlotSet` and `BotUttered` events.
Any other event type will be filtered out by the default action [`action_extract_slots`](./default-actions.mdx#action_extract_slots).
The name of the custom action must be specified in the `action` key of the relevant
[custom slot mapping](./domain.mdx#custom-slot-mappings) in the domain.
Note that the action name must be listed in the domain `actions` section too.

## `ValidationAction` class

> You can extend the `ValidationAction` class in the Rasa SDK to define custom extraction and / or validation of slots that
> can be set or updated outside of a form context.
> With this option, you do not need to specify the `action` key in the [custom slot mapping](./domain.mdx#custom-slot-mappings).
> 

Reviewer: There needs to be a N.B. note here that says, if you use this class to to do custom extraction, mapping conditions are ignored, in addition this suggestion since this will run even if a form is active:

```suggestion
You can extend the `ValidationAction` class in the Rasa SDK to define custom extraction and / or validation of slots that
can be set or updated whether a form is active or not.
With this option, you do not need to specify the `action` key in the [custom slot mapping](./domain.mdx#custom-slot-mappings), since the default action `action_extract_slots` runs `action_validate_slot_mappings` automatically. 

```

Author: Yup, will apply this suggestion. Do you mean bringing the existing N.B note further up after the first sentence?

Reviewer: Ja, bring it up and phrase it stronger (I can't find the comment I may or may not have left on it sorry!)

Author: Done!

Note that you do not need to implement the `name` method in the custom action extending `ValidationAction`, since this
is already provided: `action_validate_slot_mappings`. However this name must be added to the domain `actions` list.

> If you override the `name` method, the custom validation action will not run because the original name has been hardcoded

Reviewer: ```suggestion
If you override the `name` method, the custom validation action will not run because the original name is hardcoded
```

in the call that the default action `action_extract_slots` makes to the action server.

> 

Reviewer: What happens with non-rasa-sdk action servers here? Can the users implement their own `action_validate_slot_mappings` ?

Reviewer: This pertains to the last point in the PR about updating action server docs - Some of this might need to be moved over there entirely, so that people with non-sdk action servers also can refer to and understand how to use custom actions for slot mappings.

Author: > What happens with non-rasa-sdk action servers here?

Completely forgot this is also possible 🙈 I guess they'll have to implement their own `ValidationAction` mirroring the Python sdk. Should I then open a new issue for action server docs update? What kind of issues need to be updated, I made a note about knowledge base slots code that might need to be updated with  mapping of `type: custom`?

Reviewer: >Completely forgot this is also possible

And I completely neglected to consider it in this review up till now, so equally guilty 🙈

Yes, let's make a new issue for sdk docs updates. 
That issue can include 
- deciding how much of the slot validation action docs should move to the action server docs
- describing the behaviour of rasa-sdk's `ValidationAction` in broad strokes to allow other language implementations of it.

Author: In our inbox: https://github.com/RasaHQ/rasa-sdk/issues/597

        else:
            # validation failed, set this slot to None
            return {"location": None}
```

### Extraction of Custom Slot Mappings

In order to define custom extraction code, you must write an `extract_<slot_name>` method for every slot with a
custom slot mapping.

The following example shows the implementation of a custom action that extracts the slot `count_of_insults` to keep
track of the user's attitude.

```python
from typing import Dict, Text, Any

from rasa_sdk import Tracker
from rasa_sdk.executor import CollectingDispatcher
from rasa_sdk.forms import ValidationAction

class ValidateCustomSlotMappings(ValidationAction):
    async def extract_count_of_insults(
        self, dispatcher: CollectingDispatcher, tracker: Tracker, domain: Dict
    ) -> Dict[Text, Any]:
        intent_of_last_user_message = tracker.get_intent_of_latest_message()
        current_count_of_insults = tracker.get_slot("count_of_insults")
        if intent_of_last_user_message == "insult":
           current_count_of_insults += 1

        return {"count_of_insults": current_count_of_insults}
```

## `FormValidationAction` class

> If certain custom slot mappings should only be extracted and / or validated within the context of a form,
> make sure your custom action inherits from `FormValidationAction` rather than from `ValidationAction`.
> It is not required to use mapping conditions for custom slot mappings defined in a custom action extending
> `FormValidationAction`.
> Note that this custom action will only run if the form is activated.
> 
> To learn more about how to implement this class, see Forms [Advanced Usage](./forms.mdx#advanced-usage).

Reviewer: ```suggestion
A `FormValidationAction` custom action will only run if the form it specifies in its name is activated.
If certain custom slot mappings should only be extracted and / or validated within the context of a form,
the custom action should inherit from `FormValidationAction` rather than from `ValidationAction`.
Since `FormValidationAction` runs on every user turn only as long as the form is active, it is not required to use mapping conditions for custom slot mappings defined in a custom action extending
`FormValidationAction`.

To learn more about how to implement this class, see Forms [Advanced Usage](./forms.mdx#advanced-usage).
```

---

Rasa SDK provides three base classes that you can inherit from in order to implement custom slot extraction and
validation logic:

## `Action` class

You can use this class to create a regular [custom action](./custom-actions.mdx). The base class will not prevent you
from returning arbitrary events, however a slot validation action should only return `SlotSet` and `BotUttered` events.
Any other event type will be filtered out by the default action [`action_extract_slots`](./default-actions.mdx#action_extract_slots).
The name of the custom action must be specified in the `action` key of the relevant
[custom slot mapping](./domain.mdx#custom-slot-mappings) in the domain.
Note that the action name must be listed in the domain `actions` section too.

## `ValidationAction` class

You can extend the `ValidationAction` class in the Rasa SDK to define custom extraction and / or validation of slots that
can be set or updated outside of a form context.

> :::note
> `ValidationAction` will ignore any slots mapped to a form added to the slot mapping's `conditions`.
> Please extend the [`FormValidationAction`](./slot-validation-actions.mdx#formvalidationaction-class) class for applying
> custom slot mappings only within the context of a form.
> :::

Reviewer: ```suggestion
:::note
`ValidationAction` is intended for extracting slots **outside**  the context of a form. It will ignore extraction and validation methods for any slots that have a form specified under the slot mapping's `conditions`. It will not run these methods when the specified form is active nor when no form is active.
Please extend the [`FormValidationAction`](./slot-validation-actions.mdx#formvalidationaction-class) class for applying
custom slot mappings only within the context of a form.
:::
```

Author: Whoop whoop so excited about merging this, yipeee 😄
